File: elixir_umbrella/apps/world/lib/perlin_noise.ex
defmodule PerlinNoise do
use Bitwise

  # returns int
  # int minY, int maxY, float t
  defp interpolate(minY, maxY, t) do
    u = 1 - t
    minY * u + maxY * t
  end

  # Return a width*height list of random floats between 0 and 1
  defp GenerateWhiteNoise(width, height) do
    #Stream.repeatedly(fn -> :random.uniform end) |> Enum.take(width * height)
    :array.new([{:size,height}, {:fixed,:true}, {:default, 
      :array.new([{:size,width}, {:fixed,:true}, {:default, :random.uniform}]))
  end

  # Returns int[][]
  # int minY, int maxY, float[][] perlinNoise
  defp MapInts(minY, maxY, perlinNoise) do
    #width = perlinNoise.Length;
    #height = perlinNoise[0].Length;
    #int[][] heightMap = GetEmptyArray<int>(width, height);
    #for (int i = 0; i < width; i++)
    #  for (int j = 0; j < height; j++)
    #    heightMap[i][j] = Interpolate(minY, maxY, perlinNoise[i][j]);
    #}
    #return heightMap;
    Enum.map(perlinNoise, fn {i} -> Interpolate(minY, maxY, i) end)
  end

  defp lookup(array, x, y) do
    :array.get(x, :array.get(y, array))
  end

  # returns float[][]
  # float[][] baseNoise, int octave
  defp GenerateSmoothNoise(width, height, baseNoise, octave) do
    samplePeriod = 1 <<< octave  # calculates 2 ^ k
    sampleFrequency = 1.0 / samplePeriod

    smoothNoise = :array.new([{:size,height}, {:fixed,:true}, {:default, 
      :array.new([{:size,width}, {:fixed,:true}, {:default, 0}]))

    #for (int i = 0; i < width; i++)
    Enum.each(list.seq(0, width-1), 
    fn (i) ->
      # calculate the horizontal sampling indices
      iSample0 = div(i, samplePeriod) * samplePeriod
      iSample1 = rem( (iSample0 + samplePeriod), width) #wrap around
      horizontalBlend = (i - iSample0) * sampleFrequency

      Enum.each(list.seq(0, height-1), 
      fn (j) ->
        # calculate the vertical sampling indices
        jSample0 = div(j, samplePeriod) * samplePeriod
        jSample1 = rem( (jSample0 + samplePeriod), height) #wrap around
        verticalBlend = (j - jSample0) * sampleFrequency

        # blend the top two corners
        top = Interpolate(
          lookup(baseNoise,iSample0,jSample0), 
          lookup(baseNoise,iSample1,jSample0), 
          horizontalBlend)

        # blend the bottom two corners
        bottom = Interpolate(
          lookup(baseNoise,iSample0,jSample1), 
          lookup(baseNoise,iSample1,jSample1), 
          horizontalBlend)

        # final blend
        smoothNoise[i][j] = Interpolate(top, bottom, verticalBlend)
      )
    )
    return smoothNoise
  end


  defp Blend([p1|perlinNoise], [o1|octaveNoise], amplitude, result) do
    Blend(perlinNoise, octaveNoise, amplitude, result ++ [p1 + o1 * amplitude])
  end
  defp Blend([], [], amplitude, result) do
    result
  end

  # returns float[][] 
  # float[][] baseNoise, int octaveCount
  defp GeneratePerlinNoise(width, height, baseNoise, octaveCount) do
    smoothNoise = new float[octaveCount][][]; #an array of 2D arrays containing
    PERSISTANCE = 0.4f

    # generate smooth noise
    octaves = list.seq(octaveCount-1, 0, -1)
    smoothNoise = Enum.map(octaves, fn (octave) -> GenerateSmoothNoise(width,height, baseNoise, octave) end)

    perlinNoise = GetEmptyArray(width, height) #an array of floats initialised to 0

    amplitude = 1f
    totalAmplitude = 0.0f

    # blend noise together
    Enum.map(smoothNoise, fn (octaveNoise) -> 
      amplitude = amplitude * PERSISTANCE
      totalAmplitude = totalAmplitude + amplitude
      perlinNoise = Blend(perlinNoise, octaveNoise, amplitude, [])
    end)

    # normalisation
    normalizedPerlinNoise = Enum.map(perlinNoise, fn {i} -> i / totalAmplitude end)
  end

  #--------------------------------------------------------------------------

  def SeedRandom() do
    :random.seed(:erlang.now)
  end

  # returns int[][] 
  # int width, int height, int minY, int maxY, int octaveCount
  def GetIntMap(width, height, minY, maxY, octaveCount) do
    baseNoise = GenerateWhiteNoise(width, height)
    perlinNoise = GeneratePerlinNoise(width, height, baseNoise, octaveCount)
    MapInts(minY, maxY, perlinNoise)
  end

  #--------------------------------------------------------------------------
end

File: EclipseWorkspace/WorldServer/src/PerlinNoise.cpp
#include <iostream>
#include <stdlib.h>
#include "PerlinNoise.h"
using namespace std;

  PerlinNoise::PerlinNoise(unsigned long int seed)
  {
    //MTRand_int32 irand(init, length); // 32-bit int generator
    mRand(); // 32-bit double generator
    mRand.seed(seed);
    // this is an example of initializing by an array
    // you may use MTRand(seed) with any 32bit integer
    // as a seed for a simpler initialization
    //MTRand drand; // double in [0, 1) generator, already init

    // generate the same numbers as in the original C test program
    //    std::printf("1000 32-bit integer random numbers:\n");
    //    for (int i = 0; i < 1000; ++i) {
    //      std::printf("%10lu ", irand());
    //      if ((i % 5) == 4) std::printf("\n");
    //    }
    //    std::printf("\n1000 random numbers in [0, 1):\n");
    //    for (int i = 0; i < 1000; ++i) {
    //      std::printf("%10.8f ", drand());
    //      if ((i % 5) == 4) std::printf("\n");
    //    }
  }

  TArray2<float>* PerlinNoise::GenerateWhiteNoise(uint width, uint height)
  {
    TArray2<float>* noise = new TArray2<float>(width,height);
    for (uint i = 0; i < width; i++)
    {
      for (uint j = 0; j < height; j++)
      {
        (*noise)(i,j) = mRand();
      }
    }
    return noise;
  }

  float PerlinNoise::Interpolate(float minY, float maxY, float t)
  {
    float u = 1 - t;
    return (minY * u + maxY * t);
  }
  uint PerlinNoise::Interpolate(uint minY, uint maxY, float t)
  {
    float u = 1 - t;
    return (uint)(minY * u + maxY * t);
  }

  TArray2<uint>* PerlinNoise::MapInts(uint minY, uint maxY, TArray2<float>* perlinNoise)
  {
    TArray2<uint>* heightMap = new TArray2<uint>(perlinNoise->rows(), perlinNoise->cols());
    for (uint i = 0; i < perlinNoise->rows(); i++)
    {
      for (uint j = 0; j < perlinNoise->cols(); j++)
      {
        (*heightMap)(i,j) = Interpolate(minY, maxY, (*perlinNoise)(i,j));
      }
    }
    return heightMap;
  }

  TArray2<float>* PerlinNoise::GenerateSmoothNoise(int octave, TArray2<float>* baseNoise)
  {
    TArray2<float>* smoothNoise = new TArray2<float>(baseNoise->rows(), baseNoise->cols());
    uint samplePeriod = 1 << octave; // calculates 2 ^ k
    float sampleFrequency = 1.0f / samplePeriod;

    for (uint i = 0; i < baseNoise->rows(); i++)
    {
      //calculate the horizontal sampling indices
      uint iSample0 = (i / samplePeriod) * samplePeriod;
      uint iSample1 = (iSample0 + samplePeriod) % baseNoise->rows(); //wrap around
      float horizontalBlend = (i - iSample0) * sampleFrequency;

      for (uint j = 0; j < baseNoise->cols(); j++)
      {
        //calculate the vertical sampling indices
        uint jSample0 = (j / samplePeriod) * samplePeriod;
        uint jSample1 = (jSample0 + samplePeriod) % baseNoise->cols(); //wrap around
        float verticalBlend = (j - jSample0) * sampleFrequency;

        //blend the top two corners
        float top = Interpolate((*baseNoise)(iSample0,jSample0), (*baseNoise)(iSample1,jSample0), horizontalBlend);

        //blend the bottom two corners
        float bottom = Interpolate((*baseNoise)(iSample0,jSample1), (*baseNoise)(iSample1,jSample1), horizontalBlend);

        //final blend
        (*smoothNoise)(i,j) = Interpolate(top, bottom, verticalBlend);
      }
    }

    return smoothNoise;
  }

  TArray2<float>* PerlinNoise::GeneratePerlinNoise(int octaveCount, TArray2<float>* baseNoise)
  {
    TArray2<float>* smoothNoise[octaveCount];
    const float PERSISTANCE = 0.4f;

    //generate smooth noise
    for (int octave = 0; octave < octaveCount; octave++)
    {
      smoothNoise[octave] = GenerateSmoothNoise(octave, baseNoise);
    }

    TArray2<float>* perlinNoise = new TArray2<float>(baseNoise->rows(),baseNoise->cols());
    for (uint i = 0; i < baseNoise->rows(); i++)
    {
      for (uint j = 0; j < baseNoise->cols(); j++)
      {
        (*perlinNoise)(i,j) = 0;
      }
    }

    float amplitude = 1.0f;
    float totalAmplitude = 0.0f;

    //blend noise together
    for (int octave = octaveCount - 1; octave >= 0; octave--)
    {
      amplitude *= PERSISTANCE;
      totalAmplitude += amplitude;

      for (uint i = 0; i < baseNoise->rows(); i++)
      {
        for (uint j = 0; j < baseNoise->cols(); j++)
        {
          (*perlinNoise)(i,j) += (*smoothNoise[octave])(i,j) * amplitude;
        }
      }
    }

    //normalisation
    for (uint i = 0; i < baseNoise->rows(); i++)
    {
      for (uint j = 0; j < baseNoise->cols(); j++)
      {
        (*perlinNoise)(i,j) /= totalAmplitude;
      }
    }
    return perlinNoise;
  }

  TArray2<uint>* PerlinNoise::GetIntMap(uint width, uint height, uint minY, uint maxY, uint octaveCount)
  {
    TArray2<float>* baseNoise = GenerateWhiteNoise(width, height);
    TArray2<float>* perlinNoise = GeneratePerlinNoise(octaveCount, baseNoise);
    return MapInts(minY, maxY, perlinNoise);
  }


